### PR TITLE
2312-2

### PR DESCRIPTION
**Pull request by lstpsche**
## Номер
2312

## Номер задания
2

## Ссылка на видео с демо
https://www.youtube.com/watch?v=Xbt8ltPI-GE

## Комментарии
Level 1 working. Spent all the time for fixing hound errors, and because of it didn't add Level 2.

### CODE REVIEW

**Comment by alexshgov** [comment]
I don't want my comments to hide when you change the code, that's why all general comments (ideas, useful links, etc) I will write down here.

Let's start with your class `BattleAnalyzer`.  Just looking at this name I can tell you that you use OOP badly!
In general, you should think about the business **entities** when doing OOP.
I see that this class can be just a `Battle`.

Then, think about **messages** which you will pass to the `Battle` object.
Consider the following code.

```ruby
class Battle
  attr_reader :battle_file_path, ... #etc
  
  def intialize(battle_file_path)
     # ...
  end

  def words_count; end
  def rounds_count; end
  def bad_words_count; end
  def battler_name; end
  # ...
end
```
I guess you've got the idea now.

Then let's talk about the `VersusAnalyzer` class. To think about this class let's answer the question: "What is the input and what is the output of the program?"

Obviously, the input is our files, and the output is...table printed with battlers info! Right?!

**Think about your code like user, not a developer. Think about API which you expose!**

Basically, I just want my `BattlersTable` to be printed! Not runned, not executed!
```ruby
BattlersTable.new(args = {}).print
```
I do not care what is under the hood for now. I'm just thinking about API.


Think about these suggestions, thanks.

Useful links:
* http://www.carlopescio.com/2011/04/your-coding-conventions-are-hurting-you.html

File: 2312/2/battle.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  end
+
+  def rounds_amount(counter = 0)
+    IO.read(@battle_file_path).each_line { |line| counter += 1 if line[/Раунд \d/] }
+    counter
+  end
+
+  def bad_words_amount(rapper)
+    counter = 0
+    all_words(rapper).each { |word| counter += 1 if word.include?('*') || RussianObscenity.obscene?(word) }
```
Comment (alexshgov) [inline 2312/2/battle.rb]:
```ruby
all_words_said_by(rapper_name).select { |word| word.include?('*') || RussianObscenity.obscene? }.count
```
Even more better to introduce `Word` class which will respond to the `#obscene?` message.
Then the code will look like this:
```ruby
all_words_said_by(rapper_name) # returns an array of Word class objects

all_words_said_by(rapper_name).select(&:obscene?).count # returns obscene words
```

Do not forget about useful Enumerable methods!

File: 2312/2/battle.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    if paired?
+      lyrics_double(rapper).split(/[.,!?:;-]/).join(' ')
+    else
+      lyrics_single.split(/[.,!?:;-]/).join(' ')
+    end
+  end
+
+  def lyrics_single
+    lyrics = ''
+    File.open(@battle_file_path, 'r').each_line { |line| lyrics += ' ' + line.strip }
```
Comment (alexshgov) [inline 2312/2/battle.rb]: What is the intention of performing `strip`?

File: 2312/2/battle.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    lyrics
+  end
+
+  def lyrics_double(rapper)
+    lyrics = ''
+    File.open(@battle_file_path, 'r').each_line do |line|
+      line_analyzer = Line.new(line)
+      lyrics = line_analyzer.writable_line(rapper)
```
Comment (alexshgov) [inline 2312/2/battle.rb]: What does `writable_line` mean?

Anyway, consider the following code:

```ruby
File.open(@battle_file_path, 'r').map do |line_text|
  line = Line.new(line_text)
  next unless line.belongs_to?(rapper)
  line.apply_formatting! # this should perform all of your magic strips 
end.join(' ')
```
Here I use `#map` instead of your lyrics hack.
Moreover, I assume that `Line` object will respond to `#belongs_to?` message.
Then I expect to have specific method `#apply_formatting!` to format the line.

Comment (alexshgov) [reply 2312/2/battle.rb]: Note: Rubocop will complain about chaining methods, but it is the problem for future changes.

File: 2312/2/battlers_table.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+require 'terminal-table'
+
+# BattlersTable prints a table of data, you provide to it
+class BattlersTable
+  def initialize(battlers, num_top)
+    @battlers = battlers
+    @counter = -1
+    @table = []
+    @table = create_table(num_top)
```
Comment (alexshgov) [inline 2312/2/battlers_table.rb]: Ideally, you should not place any logic in the initializer.
It should set initial state only.

You allowed to do this preparation in `#print` method if needed.

```ruby
def print
  create_table
  do_something_useful

  puts Terminal::Table.new rows: table
end
```

File: 2312/2/battlers_table.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+  def create_table(num_top)
+    @battlers.sort_by { |_name, info| info.stats.words_info[:bad_words_num] }.reverse.to_h.keys.each do |name|
+      @counter += 1
+      fill_table_global(name)
+    end
+    @table[0...num_top]
+  end
+
+  def fill_table_global(name)
```
Comment (alexshgov) [inline 2312/2/battlers_table.rb]: Too hard to understand abstractions behind the `table_global` and `table_second`.
We have **only one table** with data in it.

File: 2312/2/line.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
+# this class analyzes provided line and returns a writeable one
+class Line
+  attr_reader :line
+
+  def initialize(line)
+    @line = line
+  end
+
+  def in_separate_words
```
Comment (alexshgov) [inline 2312/2/line.rb]: Just `words`. Look how great it will be read then
```ruby
line = Line.new(line_tet)
words = line.words
```
Cool, right?

File: 2312/2/line.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    line.split.map { |word| Word.new(word) }
+  end
+
+  def writable_line(rapper)
```
Comment (alexshgov) [inline 2312/2/line.rb]: Where is the method used? I can't find any occurrences.

File: 2312/2/lyrics.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      line_text.strip
+    end.join(' ')
+  end
+
+  def line_belongs_to?(rapper_name, line)
+    line_includes_rapper_name = line.include?("#{rapper_name}:")
+    if line.start_with?(/\w+:/) && !line_includes_rapper_name
+      @line_belongs_to_rapper = false
+    elsif line_includes_rapper_name || @line_belongs_to_rapper
+      @line_belongs_to_rapper = true
```
Comment (alexshgov) [inline 2312/2/lyrics.rb]: **Do not mutate state!**
I believe you could rewrite your code without using the `line_belongs_to_rapper` instance variable which you mutate.

File: 2312/2/rapper.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
+require './word'
+require './lyrics'
+
+# this class provides all info about rapper
+class Rapper
+  attr_reader :stats, :name, :overall, :battles
+
+  def initialize(name, all_battles_paths)
+    @name = name
+    @battles = all_battles_paths.select { |battle_path| battle_path.split(/( против | vs )/i).first.include?(@name) }
```
Comment (alexshgov) [inline 2312/2/rapper.rb]: No logic in initializers.

File: 2312/2/rapper.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    @overall = { battles_num: @battles.size, rounds_num: Rapper.rapper_rounds(@battles) }
+    @stats = Stats.new(self)
+  end
+
+  def self.count_rounds(battle_file_path)
```
Comment (alexshgov) [inline 2312/2/rapper.rb]: Why don't you want to create Rapper object? Then we could you something like
```ruby
rapper = Rapper.new(args)
rapper.rounds_count
rapper.bad_words
```
And so on.

File: 2312/2/rapper.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    num_of_rounds = 0
+    battles.each do |battle_file_path|
+      num_of_rounds += Rapper.count_rounds(battle_file_path)
+    end
+    num_of_rounds.zero? ? 1 : num_of_rounds
+  end
+end
+
+# this class provides all stats about rapper
+class Stats < Rapper
```
Comment (alexshgov) [inline 2312/2/rapper.rb]: Wrong usage of inheritance. This is confusing that stats inherit from a rapper.

File: 2312/2/word.rb [hunk 12]
```diff
@@ -0,0 +1,8 @@
+require 'russian_obscenity'
+
+# this class has only one function for now - to say if the word is obscene
+class Word
+  def self.obscene?(word)
```
Comment (alexshgov) [inline 2312/2/word.rb]: Why self?!
You should create the `Word` object and then **send** `#obscene?` **message** to it!

File: 2312/2/battle.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
+require './word'
+require './lyrics'
+
+# this class analyzes text of battle
+class Battle
+  def initialize(battle_file_path)
+    @battle_file_path = battle_file_path
+  end
+
+  def extract_battlers_names
```
Comment (alexshgov) [inline 2312/2/battle.rb]: Again, think about business entities when writing OOP code.
Medidate on this code.

`battle#battlers_names` vs `battle#extract_battles_names`
`battle#rounds_count` vs `battle#count_rounds`

File: 2312/2/battlers_table.rb [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+class BattlersTable
+  def initialize(battlers, num_top)
+    @battlers = battlers
+    @counter = -1
+    @table = []
+    @num_top = num_top
+  end
+
+  def print
+    if @num_top.zero?
```
Comment (alexshgov) [inline 2312/2/battlers_table.rb]: Avoid using `@var`. Add `attr_reader :var` and use it instead.

Comment (alexshgov) [inline 2312/2/battlers_table.rb]: Let's rewrite it a little bit.
```ruby
def print
   terminate_with(message: num_top_zero_message) if num_top.zero?
  
   create_table! # we use bang! because it modifies the object
   puts Terminal::Table.new rows: table[0...num_top]
end
```

File: 2312/2/lyrics.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  private
+
+  def lyrics_solo_battle
+    File.open(@battle_file_path, 'r').select { |line_text| line_text }.join(' ')
+  end
+
+  def lyrics_in_paired_battle_by(rapper_name)
+    line_belongs = false
+    File.open(@battle_file_path, 'r').select do |line_text|
+      line_belongs = Line.new(line_text).belongs_to?(rapper_name, line_belongs)
```
Comment (alexshgov) [inline 2312/2/lyrics.rb]: What is going here?!
Why do you mutate `line_belongs` variable??

I don't expect `Line#belongs_to?` to be dependent on this weird state argument.
You should stick to `Line.new(line_text).belongs_to?(rapper_name)` interface.

File: 2312/2/rapper.rb [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    @overall[:rounds_num] = rounds.zero? ? 1 : rounds
+  end
+
+  def rapper_rounds
+    @battles.map { |battle_file_path| Battle.new(battle_file_path).count_rounds }.count
+  end
+end
+
+# this class provides all stats about rapper
+class Stats
```
Comment (alexshgov) [inline 2312/2/rapper.rb]: One file -- one class.

If this is a rapper stats, move it to the appropreate namespace.
```ruby
module Rapper
   class Stats; end
end
```

File: 2312/2/word.rb [hunk 17]
```diff
@@ -0,0 +1,12 @@
+require 'russian_obscenity'
+
+# this class has only one function for now - to say if the word is obscene
+class Word
+  def initialize(word)
+    @word = word
+  end
+
+  def obscene?
+    @word.include?('*') || RussianObscenity.obscene?(@word)
```
Comment (alexshgov) [inline 2312/2/word.rb]: Minor thing: add  `attr_reader :word` and don't use `@word` variable

File: 2312/2/versus.rb [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    collect_all_battles_files
+    collect_battlers
+  end
+
+  def run
+    top_bad(@to_do[:top_bad_words])
+    # top_words(@to_do[:top_words], @to_do[:name])
+  end
+
+  # def top_words(num_top, rapper_name)
```
Comment (alexshgov) [inline 2312/2/versus.rb]: Commented code?!

**Comment by alexshgov** [comment]
Keep going!

File: 2312/2/battlers_table.rb [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  private
+
+  def terminate_with(message)
+    puts message
+    exit
+  end
+
+  def create_table; end
```
Comment (alexshgov) [inline 2312/2/battlers_table.rb]: Empty method?

File: 2312/2/battlers_table.rb [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+  def create_table!
+    @battlers.sort_by { |_name, info| info.stats.rapper_stats[:bad_words_num] }.reverse.to_h.keys.each do |name|
+      @counter += 1
```
Comment (alexshgov) [inline 2312/2/battlers_table.rb]: I've already told you not to mutate this counters

File: 2312/2/command_parser.rb [hunk 21]
```diff
@@ -0,0 +1,52 @@
+# this class here only because of reek.
```
Comment (alexshgov) [inline 2312/2/command_parser.rb]: ???

Comment (alexshgov) [reply 2312/2/command_parser.rb]: Your classes should be created for purpose

File: 2312/2/line.rb [hunk 22]
```diff
@@ -0,0 +1,15 @@
+# this class shows belonging of line of lyrics to rapper
+class Line
+  def initialize(line)
+    @line = line
+  end
+
+  def belongs_to?(rapper_name, line_belongs)
```
Comment (alexshgov) [inline 2312/2/line.rb]: You ignored my words about `line_belongs` argument.

File: 2312/2/rapper.rb [hunk 23]
```diff
@@ -0,0 +1,63 @@
+require './word'
+require './lyrics'
+
+# this class provides all info about rapper
+class Rapper
+  # this class provides all stats about rapper
+  class Stats
```
Comment (alexshgov) [inline 2312/2/rapper.rb]: Move this class to a separate file.
